crates/red_knot_python_semantic/resources/mdtest/import/conventions.md

```diff
@@ -0,0 +1,474 @@
+__all__ = ['Foo']
+```
+
+`b.pyi`:
+
+```pyi
+class Foo: ...
+```
+
+## Runtime files
+
+For runtime files (i.e., `.py` files), the re-exported symbols should be inferred correctly.
+
+TODO: When we support rule selection, the `implicit-reexport` rule should be disabled by default and
+this test case should be updated to explicitly enable it.
+
+```py
+# error: "Module `b` does not explicitly export attribute `Foo`"
+from b import Foo
+
+reveal_type(Foo)  # revealed: Literal[Foo]
+```
+
+`b.py`:
+
+```py
+from c import Foo
+```
+
+`c.py`:
+
+```py
+class Foo: ...
+```
+
+## Non-exports in `__init__.py`
+
+Red knot does not special case `__init__.py` files, so if a symbol is imported in `__init__.py`
+which is not re-exported, it should raise an error.
+
+TODO: When we support rule selection, the `implicit-reexport` rule should be disabled by default and
+this test case should be updated to explicitly enable it.
+
+```py
+# error: 15 "Module `a` does not explicitly export attribute `Foo`"
+# error: 20 "Module `a` does not explicitly export attribute `c`"
+from a import Foo, c, foo
+
+reveal_type(Foo)  # revealed: Literal[Foo]
+reveal_type(c)  # revealed: <module 'a.b.c'>
+reveal_type(foo)  # revealed: <module 'a.foo'>
+```
+
+`a/__init__.py`:
+
+```py
+from .b import c
+from .foo import Foo
+```
+
+`a/foo.py`:
+
+```py
+class Foo: ...
+```
+
+`a/b/__init__.py`:
+
+```py
+```
+
+`a/b/c.py`:
+
+```py
+```
+
+## Re-exports in `__init__.py`
+
+But, if they're re-exported, it should not raise an error.
+
+TODO: When we support rule selection, the `implicit-reexport` rule should be disabled by default and
+this test case should be updated to explicitly enable it.
+
+```py
+from a import Foo, c, foo
+
+reveal_type(Foo)  # revealed: Literal[Foo]
+reveal_type(c)  # revealed: <module 'a.b.c'>
+reveal_type(foo)  # revealed: <module 'a.foo'>
+```
+
+`a/__init__.py`:
+
+```py
+from .b import c as c
+from .foo import Foo as Foo
+```
+
+`a/foo.py`:
+
+```py
+class Foo: ...
+```
+
+`a/b/__init__.py`:
+
+```py
+```
+
+`a/b/c.py`:
+
+```py
+```
+
+## Re-exports in `__init__.pyi`
+
+Similarly, for an `__init__.pyi` (stub) file, importing a non-exported name should raise an error
+but the inference would be `Unknown`.
+
+```py
+# error: 15 "Module `a` has no member `Foo`"
+# error: 20 "Module `a` has no member `c`"
+from a import Foo, c, foo
+
+reveal_type(Foo)  # revealed: Unknown
+reveal_type(c)  # revealed: Unknown
+reveal_type(foo)  # revealed: <module 'a.foo'>
+```
+
+`a/__init__.pyi`:
+
+```pyi
+from .b import c
+from .foo import Foo
+```
+
+`a/foo.pyi`:
+
+```pyi
+class Foo: ...
+```
+
+`a/b/__init__.pyi`:
+
+```pyi
+```
+
+`a/b/c.pyi`:
+
+```pyi
+```
+
+## Conditional re-export in stub file
+
+The following scenarios are when a re-export happens conditionally in a stub file.
+
+### Global import
+
+```py
+# error: [possibly-unbound-import] "Member `Foo` of module `a` is possibly unbound"
+from a import Foo
+
+reveal_type(Foo)  # revealed: Literal[Foo] | str
```

Comment:
This should be just `str`; the `Foo` class type should not be visible here at all. I think this behavior change will naturally fall out of the alternate implementation strategy.